synapse/storage/database.py

```diff
@@ -507,8 +506,9 @@ def new_transaction(
             self._txn_perf_counters.update(desc, duration)
             sql_txn_timer.labels(desc).observe(duration)
 
-    @defer.inlineCallbacks
-    def runInteraction(self, desc: str, func: Callable, *args: Any, **kwargs: Any):
+    async def runInteraction(
+        self, desc: str, func: Callable, *args: Any, **kwargs: Any
+    ) -> Any:
```

Comment:
I think this can be:

```suggestion
    async def runInteraction(
        self, desc: str, func: "Callable[..., R]", *args: Any, **kwargs: Any
    ) -> R:
```

But mypy then complains:

> synapse/storage/database.py:550: error: Incompatible return value type (got "R", expected "R")  [return-value]

I suspect something isn't getting piped through `runWithConnection` / `new_transaction` properly?

Reply:
Does the `inner_func` in `runWithConnection` need annotating as returning `R`?

Reply:
So I think the disconnect is happening because mypy doesn't know that the return type of `self.new_transaction` is the same `R` as returned by `func`, due to the indirection of `runWithConnection`.

The signature `runWithConnection(func, *args, **kwargs)` means that when we call it with `runWithConnection(self.new_transaction, func)` mypy only sees us calling `self.new_transaction(*args)` as the typing of `func` is lost due to the variadic parameters.

Changing it to `runWithConnection(lambda conn: self.new_transaction(conn, func))` works, but it may be simpler to do a `result = cast(R, result)` instead.

Reply:
Woot! Thanks. See ca31cfeb5572d3abda2790ccdc9995b79502dd0a for this.